src/FondOfSpryker/Zed/TaxProductConnector/Business/Calculator/ProductItemTaxRateByRegionCalculator.php

```diff
@@ -0,0 +1,316 @@
+
+    /**
+     * @var float
+     */
+    protected $defaultTaxRate;
+
+    /**
+     * @var null|int
+     */
+    protected $defaultRegionId;
+
+    /**
+     * @param \Spryker\Zed\TaxProductConnector\Persistence\TaxProductConnectorQueryContainerInterface $taxQueryContainer
+     * @param \Spryker\Zed\TaxProductConnector\Dependency\Facade\TaxProductConnectorToTaxInterface $taxFacade
+     */
+    public function __construct(
+        TaxProductConnectorQueryContainerInterface $taxQueryContainer,
+        TaxProductConnectorToTaxInterface $taxFacade
+    ) {
+        $this->taxQueryContainer = $taxQueryContainer;
+        $this->taxFacade = $taxFacade;
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return void
+     */
+    public function recalculate(QuoteTransfer $quoteTransfer)
+    {
+        $itemTransfers = $this->recalculateWithItemTransfers($quoteTransfer->getItems());
+        $quoteTransfer->setItems($itemTransfers);
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\CalculableObjectTransfer $calculableObjectTransfer
+     *
+     * @return \Generated\Shared\Transfer\CalculableObjectTransfer
+     */
+    public function recalculateWithCalculableObject(CalculableObjectTransfer $calculableObjectTransfer): CalculableObjectTransfer
+    {
+        $itemTransfers = $this->recalculateWithItemTransfers($calculableObjectTransfer->getItems());
+        $calculableObjectTransfer->setItems($itemTransfers);
+
+        return $calculableObjectTransfer;
+    }
+
+    /**
+     * @param \ArrayObject|\Generated\Shared\Transfer\ItemTransfer[] $itemTransfers
+     *
+     * @return \ArrayObject|\Generated\Shared\Transfer\ItemTransfer[]
+     */
+    protected function recalculateWithItemTransfers(ArrayObject $itemTransfers): ArrayObject
+    {
+
+        $foundResults = $this->getTaxSet($itemTransfers);
+
+        $taxRatesByIdProductAbstractAndCountry = $this->mapByIdProductAbstractAndCountry($foundResults);
+
+        foreach ($itemTransfers as $itemTransfer) {
+            $taxRate = $this->getEffectiveTaxRate(
+                $taxRatesByIdProductAbstractAndCountry,
+                $itemTransfer->getIdProductAbstract(),
+                $this->getShippingCountryIso2CodeByItem($itemTransfer)
+            );
+            $itemTransfer->setTaxRate($taxRate);
+        }
+
+        return $itemTransfers;
+    }
+
+    /**
+     * @param \ArrayObject $itemTransfers
+     * @return \Orm\Zed\Tax\Persistence\SpyTaxSet[]|\Propel\Runtime\Collection\ObjectCollection
+     */
+    protected function getTaxSet(ArrayObject $itemTransfers)
+    {
+        $idRegions = $this->getRegionIds($itemTransfers);
+        $idProductAbstracts = $this->getIdProductAbstract($itemTransfers);
+        $countryIso2Code = $this->getCountryIso2Codes($itemTransfers);
+        // Fallback to country if no regions are found
+        if (count($idRegions) === 0) {
+            return $this->taxQueryContainer
```

Comment:
It would be nicer with repository instead of querycontainer

Reply:
True.. wasn't sure since the original design uses a query container as well.. I'll change that and will call the query container in the repository